test/models/announcement_test.exs

```diff
@@ -129,5 +129,19 @@ defmodule Constable.AnnouncementTest do
       assert announcement_with_user_comment.id in announcement_ids
       refute announcement_without_user_comment.id in announcement_ids
     end
+
+    test "includes each announcement once, even with multiple comments by user" do
+      user = insert(:user)
+      announcement = insert(:announcement)
+
+      insert_pair(:comment, announcement: announcement, user: user)
+
+      found_announcement =
+        user.id
+        |> Announcement.with_comments_by_user()
+        |> Repo.one()
```

Comment:
It has been a while since Ive worked with Ecto, but doesnt `Repo.one` guarantee only one record is returned? Would this better protect against a false positive if you did a `Repo.all` and then asserted that only one record was returned in the result?

If not, could you show me where my understanding of `Repo.one` is off, or my understanding of this spec is off? Thanks :)

Reply:
Yeah, I always find `Repo.one` a bit confusing. 

`Repo.one` will _raise_ if more than one value is being returned. So it's not doing something like a `limit 1` (that would be [Repo.first/1](https://hexdocs.pm/ecto/Ecto.Query.html#first/2)). Instead, it's meant as a function that guarantees one or none to be returned. 

From [the Repo.one docs](https://hexdocs.pm/ecto/Ecto.Repo.html#c:one/2): 

> Fetches a single result from the query. Returns nil if no result was found. Raises if more than one entry.

The [`!` version of it](https://hexdocs.pm/ecto/Ecto.Repo.html#c:one!/2) will additionally raise if no records are found. 

Does that make more sense?

Reply:
Ah, my eyes glazed right over that `raises if...` part! TYSM for highlighting that. This makes total sense now, and is a nice guarantee that the test always produces a true positive or negative.